lab/src/mapreduce/schedule.go

```diff
@@ -30,5 +33,42 @@ func schedule(jobName string, mapFiles []string, nReduce int, phase jobPhase, re
 	//
 	// Your code here (Part III, Part IV).
 	//
+
+	var workersPool sync.WaitGroup
+	var taskLock sync.Mutex
+
+	undoTask := ntasks
+	taskRemain := make(chan int, ntasks)
+	for taskNum := 0; taskNum < ntasks; taskNum++ {
+		taskRemain <- taskNum
+	}
+
+	for currentTaskNum := range taskRemain {
+		worker := <- registerChan
+		workersPool.Add(1)
```

Comment:
Looks like you're using a `WaitGroup` to ensure that only one goroutine can run at the same time? May I know why we don't want to run multiple goroutines concurrently?

Reply:
WaitGroup will allow a collection of goroutine run at same time, and we call Add() to increase the counter inside the watiGroup, when we call wait(), waitGroup will block when the counter inside is not 0, by calling Done() could decrease the counter
Check this: https://golang.org/pkg/sync/#WaitGroup